oracle/src/tests.rs

```diff
@@ -161,6 +161,42 @@ fn should_combined_data() {
 	});
 }
 
+#[test]
+fn should_recombine_after_expiry() {
+	new_test_ext().execute_with(|| {
+		let key: u32 = 50;
+		let expiration_period = 1000;
+
+		OracleMembers::set(vec![10, 11, 12, 13, 14]);
+		ExpiresIn::set(expiration_period);
+
+		// submit 1000, 1100, 1200, 1300, 1400
+		for idx in 0..5 {
+			Timestamp::set_timestamp(idx * 100);
+			assert_ok!(ModuleOracle::feed_values(
+				Origin::signed(idx as u128 + 10),
+				vec![(key, idx as u32 * 100 + 1000)]
+			));
+		}
+
+		let expected = Some(TimestampedValue {
+			value: 1200,
+			timestamp: 200,
+		});
+
+		assert_eq!(ModuleOracle::get(&key), expected);
+
+		// make the first two values expire - this should set median to 1300.
+		// note that we do not submit new values to trigger re-combination
+		Timestamp::set_timestamp(150 + expiration_period);
+		let expected = Some(TimestampedValue {
+			value: 1300,
+			timestamp: 300,
+		});
+		assert_eq!(ModuleOracle::get(&key), expected);
```

Comment:
Can this ever return an `Err`? What happens if all past entries are stale?

Reply:
`get` is returning an `Option`, not a `Result`. By default, it will only return None until the first valid result is produced. After that, if too many past entries get stale, the latest valid aggregate is returned. After this PR gets merged, you can check the timestamp of the returned value - if it is expired, it is not valid. Note that this wasn't true before PR, because the aggregate could include expired values, even if there were sufficient unexpired ones.

It is possible to provide a custom `combine_data` function that returns `None` when there are insufficient valid values, instead of the latest valid one. However, this would currently cause the results to not get cached, since `None` is treated as an error, rather than a valid return result. Maybe `combine_data` should return `Result<(Option<TimestampedValue>, Option<Moment>), ())` ? Then `Err(())` would indicate that an error occurred in `combine_ok`, and that the result should not be stored, whereas `Ok((None, _))` would indicate the calculation was successful, so the result should be cached, but the value to be returned from `get` should be `None`. The error type could maybe also be generic rather than `()`, but since we'd be neither propagating it, nor printing it, I don't know if that's useful.

Alternatively, instead of the current `Option<(TimestampedValue, Option<Moment>)>`, we could return `(Option<TimestampedValue>, Option<TimestampedValue>)`, and then _always_ store the result in storage, before returning it.

@xlc @gregdhill what do you think?

Reply:
`Result<(Option<TimestampedValue>, Option<Moment>), ())` just way to complicated.
Another way is return an enum.

Reply:
An enum like this?

```rust
enum AggregateResult {
    PermanentlyInvalid(),
    PermanentlyValid(value),
    InvalidUntil(timestamp)
    ValidUntil(value, timestamp),
}
```

I would probably prefer to go with the `(Option<TimestampedValue>, Option<Moment>)` return type then, but the enum solution would work just fine as well. Let me know which one you prefer and I'll implement it.

Reply:
I implemented the enum. It makes for a somewhat uglier code, but the interface for the user is clearer. @xlc let me know what you think